certora/specs/RevertsERC20.spec

```diff
@@ -0,0 +1,52 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+methods {
+    function totalSupply() external returns uint256 envfree;
+    function balanceOf(address) external returns uint256 envfree;
+    function allowance(address, address) external returns uint256 envfree;
+    function delegatee(address) external returns address envfree;
+    function delegatedVotingPower(address) external returns uint256 envfree;
+}
+
+// Check the revert conditions for the transfer function.
+rule transferRevertConditions(env e, address to, uint256 amount) {
+    uint256 balanceOfSenderBefore = balanceOf(e.msg.sender);
+    uint256 senderVotingPowerBefore = delegatedVotingPower(delegatee(e.msg.sender));
+    uint256 recipientVotingPowerBefore = delegatedVotingPower(delegatee(to));
+
+    // Safe require because a holder's balance is added to the delegatee's voting power upon delegation.
+    require delegatee(e.msg.sender) != 0 => senderVotingPowerBefore >= balanceOfSenderBefore;
+    // Safe require because if delegatees are different the recipient's voting power excludes the holder's balance.
+    require delegatee(to) != delegatee(e.msg.sender) => recipientVotingPowerBefore <= totalSupply() - balanceOfSenderBefore;
```

Comment:
Another example of how we could refactor this:
- the first require comes from `delegatedLTEqDelegateeVP`. Note also that the condition `delegatee(e.msg.sender) != 0` is not necessary
- instead of the second require, we could have something stronger. A property similar to `twoBalancesCannotExceedTotalSupply` that we already proved: `delegatee(to) != delegatee(e.msg.sender) => recipientVotingPowerBefore + senderVotingPowerBefore <= totalSupply()`.

Reply:
I think that actually being able to prove
```
delegatee(to) != delegatee(e.msg.sender) => recipientVotingPowerBefore + senderVotingPowerBefore <= sumOfVotingPower
```
this could be harder than expected. For that, we would have to use the partial sum idea to count delegated votes. A point that could be difficulty could arise with the zero address situation.

Reply:
> this could be harder than expected. For that, we would have to use the partial sum idea to count delegated votes. A difficulty could arise with the zero address situation and the fact that we don't have totalVotingPower value.

I don't understand the difficulty: I don't think we need another `totalVotingPower` value other than the one we already have, `sumOfVotingPower`. This is assuming that we use the partial sum idea to define `sumOfVotingPower`. The zero address would not be an issue either: just take it as is. This would mean that `sumOfVotingPower` would be the sum of the actual voting powers as defined in the contract, which is what we need here. Indeed, both `recipientVotingPowerBefore` and `senderVotingPowerBefore` use the actual voting power (not the virtual one in case of the zero address)

Reply:
I'm willing to give this a try but I think it's a non trivial change that can be tackled in a separate PR (to ease review), what do you think?

Reply:
I've made a quick and dirty implementation, it seems to be not that easy, see [run](https://prover.certora.com/output/129828/6dafccd62119491e9b02564c87dc3a9c?anonymousKey=e47e5c8060e32c3bb879a995d21f9cb62aa1f2f6). I think it's related to the issue that I already discussed the Certora team.

Reply:
Good idea ! I think in the meanwhile, in this PR, we could rewrite the requires to "use that invariant". Here I mean that for all requires we should have a matching invariant in mind that it would refer to, not necesseraly needing them to be proved yet. I'll flag them below

Reply:
Done in https://github.com/morpho-org/morpho-token/pull/104.